File: golang/average_of_levels_in_binary_tree.go
package golang

// Space	: O(n)
// Time	: O(n)

import (
	"golang/model"
)


func averageOfLevels(root *model.TreeNode) []float64 {
    var ans []float64
    stack :=  []*model.TreeNode{root}
    
    for len(stack) > 0 {
        val := 0
        temp := []*model.TreeNode{}
        
        for i := 0; i < len(stack); i++ {
            val += stack[i].Val
            if stack[i].Left != nil {
                temp = append(temp, stack[i].Left)
            }
            if stack[i].Right != nil {
                temp = append(temp, stack[i].Right)
            }
        }

        ans = append(ans, float64(val) / float64(len(stack)))
        
        stack = temp
    }
    
    return ans
}

File: python3/average_of_levels_in_binary_tree.py
from .model import TreeNode

"""
Space	: O(n)
Time	: O(n)
"""

class Solution:
    def averageOfLevels(self, root: TreeNode) -> List[float]:
        ans = []
        stack = [root]
        
        while stack:
            val = 0
            temp = []
            for node in stack:
                val += node.val
                if node.left:
                    temp.append(node.left)
                if node.right:
                    temp.append(node.right)
                
            ans.append(val / len(stack))
            
            stack = temp
        
        return ans
